workflow/patterns/controlflow.py

```diff
@@ -281,49 +277,56 @@ def get_list():
                 return eng.extra_data["_Iterators"][step]["cache"]
             except KeyError:
                 if callable(get_list_function):
-                    return get_list()
+                    return get_list_function()
                 elif isinstance(get_list_function, collections.Iterable):
                     return list(get_list_function)
                 else:
-                    raise TypeError("get_list_function is not a callable nor a"
-                                    " iterable")
+                    raise TypeError("get_list_function is not callable nor an"
+                                    " iterable.")
 
-        my_list_to_process = get_list()
+        list_to_process = get_list()
 
         # First time we are in this step
         if step not in eng.extra_data["_Iterators"]:
             eng.extra_data["_Iterators"][step] = {}
             # Cache list
             if cache_data:
-                eng.extra_data["_Iterators"][step]["cache"] = get_list()
+                eng.extra_data["_Iterators"][step]["cache"] = list_to_process
             # Initialize step value
             eng.extra_data["_Iterators"][step]["value"] = {
                 "ASC": 0,
-                "DSC": len(my_list_to_process) - 1}[order]
+                "DSC": len(list_to_process) - 1}[order]
             # Store previous data
-            if 'current_data' in eng.extra_data["_Iterators"][step]:
-                eng.extra_data["_Iterators"][step]["previous_data"] = \
-                    eng.extra_data["_Iterators"][step]["current_data"]
+
+        elif 'current_data' in eng.extra_data["_Iterators"][step]:
+            eng.extra_data["_Iterators"][step]["previous_data"] = \
+                eng.extra_data["_Iterators"][step]["current_data"]
 
         # Increment or decrement step value
         step_value = eng.extra_data["_Iterators"][step]["value"]
         currently_within_list_bounds = \
-            (order == "ASC" and step_value < len(my_list_to_process)) or \
+            (order == "ASC" and step_value < len(list_to_process)) or \
             (order == "DSC" and step_value > -1)
         if currently_within_list_bounds:
             # Store current data for ourselves
             eng.extra_data["_Iterators"][step]["current_data"] = \
-                my_list_to_process[step_value]
+                list_to_process[step_value]
             # Store for the user
             if setter:
```

Comment:
Sure, if you have a question reading the code it means it is not clear enough and should be rework/commented, so no problem,  it is a great addition. 

Best